# Peer-graded Assignment: Build a Regression Model in Keras

## Concrete Strength Prediction - Part B

### Submitted by: Muhammad Jawad Bashir

**1. Importing Libararies & Dataset**

In [26]:
import pandas as pd
import numpy as np

In [27]:
concrete_data = pd.read_csv('https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv')
concrete_data.head()

Cement  Blast Furnace Slag  Fly Ash  Water  Superplasticizer  \
0   540.0                 0.0      0.0  162.0               2.5   
1   540.0                 0.0      0.0  162.0               2.5   
2   332.5               142.5      0.0  228.0               0.0   
3   332.5               142.5      0.0  228.0               0.0   
4   198.6               132.4      0.0  192.0               0.0   

   Coarse Aggregate  Fine Aggregate  Age  Strength  
0            1040.0           676.0   28     79.99  
1            1055.0           676.0   28     61.89  
2             932.0           594.0  270     40.27  
3             932.0           594.0  365     41.05  
4             978.4           825.5  360     44.30

In [28]:
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

**2. Separate into features & labels**

In [29]:
#Let's split the data into predictors and target
cd_columns = concrete_data.columns
predictors = concrete_data[cd_columns[cd_columns != 'Strength']]
target = concrete_data['Strength']
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

**2a: Normalizing the Dataset***

In [30]:
pred_norm = (predictors - predictors.mean()) / predictors.std()
pred_norm.head()

Cement  Blast Furnace Slag   Fly Ash     Water  Superplasticizer  \
0  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
1  2.476712           -0.856472 -0.846733 -0.916319         -0.620147   
2  0.491187            0.795140 -0.846733  2.174405         -1.038638   
3  0.491187            0.795140 -0.846733  2.174405         -1.038638   
4 -0.790075            0.678079 -0.846733  0.488555         -1.038638   

   Coarse Aggregate  Fine Aggregate       Age  
0          0.862735       -1.217079 -0.279597  
1          1.055651       -1.217079 -0.279597  
2         -0.526262       -2.239829  3.551340  
3         -0.526262       -2.239829  5.055221  
4          0.070492        0.647569  4.976069

In [31]:
inp_cols = pred_norm.shape[1]
inp_cols

8

**3. Defining Keras Regression Model**

In [32]:
import keras
from keras.models import Sequential
from keras.layers import Dense

In [33]:
def model_reg():
    model = Sequential()
    model.add(Dense(10, activation='relu', input_shape=(inp_cols,)))
    model.add(Dense(1))
    model.compile(optimizer = 'adam', loss='mean_squared_error')
    return model

**4. Running the model for 50 times with given requirements in Assignment Part-B**

In [34]:
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

In [35]:
mse_list = []
model = model_reg()
for i in range(0,50):
    X_train, X_test, y_train, y_test = train_test_split(pred_norm, target, test_size = 0.30, random_state=42)
    model.fit(X_train, y_train, epochs=50, verbose=0)
    y_pred = model.predict(X_test)
    mse = mean_squared_error(y_test,y_pred)
    mse_list.append(mse)

In [36]:
mse_list

[377.121195568449,
 153.78006727296923,
 113.90150140982165,
 85.46228669646383,
 68.76646358616388,
 58.10225960135661,
 50.50327087200443,
 46.29975962863619,
 43.697614338791,
 41.994568576235004,
 40.68519627070791,
 40.03355701447437,
 39.43507664449878,
 39.003679055380715,
 38.87172463394903,
 38.71650228996268,
 38.44947142494224,
 38.51534722713935,
 38.42690894357013,
 38.3101743135967,
 38.17749224327086,
 38.17494065515538,
 37.99142213810056,
 37.88379732629837,
 37.83863447319235,
 37.85352246617434,
 37.43571402343503,
 37.44726371280776,
 37.661083892570545,
 37.427370964009775,
 37.24947949096144,
 36.81475704771588,
 36.74505781692336,
 36.429215588790235,
 36.45507999158189,
 35.95066977261425,
 35.509592182647175,
 35.15802015654629,
 34.69352220700739,
 34.39374425626946,
 33.979703021894345,
 33.93716796292823,
 33.54117288242288,
 33.46912761462074,
 33.24302311733558,
 33.1260753347531,
 33.2328095316356,
 33.07183687989639,
 32.94791719977529,
 33.0568607928457

**5. MSE mean & Standard Deviation for Task-A**

In [37]:
print('Mean for Task-B: {}'.format(np.mean(mse_list)))
print('Standard Deviation for Task-B: {}'.format(np.std(mse_list)))

Mean for Task-B: 50.01945400226587
Standard Deviation for Task-B: 51.206979050697335


### Difference between MSEs of Part-A (unnromalized dataset) and Part-B(normalized dataset)

MSE in Part A = 108.159

MSE in Part B = 50.019

MSE decreased by 58.14 when we used the normalized dataset in Part-A.